sys/net/gnrc/sock/udp/gnrc_sock_udp.c

```diff
@@ -191,80 +252,50 @@ ssize_t sock_udp_send(sock_udp_t *sock, const void *data, size_t len,
     /* cppcheck-suppress nullPointer */
     if ((sock == NULL) || (sock->local.family == AF_UNSPEC)) {
         /* no sock or sock currently unbound */
-        while (src_port == 0) {
-            src_port = (uint16_t)(random_uint32() & UINT16_MAX);
-#ifdef MODULE_GNRC_SOCK_CHECK_REUSE
-            if ((sock == NULL) || !(sock->flags & SOCK_FLAGS_REUSE_EP)) {
-                /* check if port already registered somewhere */
-                for (sock_udp_t *ptr = _udp_socks; ptr != NULL;
-                     ptr = (sock_udp_t *)ptr->reg.next) {
-                    bool spec_addr = false;
-                    for (unsigned i = 0; i < sizeof(ptr->local.addr); i++) {
-                        const uint8_t *const p = (uint8_t *)&ptr->local.addr;
-                        if (p[i] != 0) {
-                            spec_addr = true;
-                        }
-                    }
-                    if (spec_addr) {
-                        continue;
-                    }
-                    if (ptr->local.port == src_port) {
-                        /* we already have one of this port registered
-                         * => generate a new one */
-                        src_port = 0;
-                    }
-                }
-            }
-#endif
-        }
         memset(&local, 0, sizeof(local));
+        if ((src_port = _get_dyn_port(sock)) == GNRC_SOCK_DYN_PORTRANGE_ERR) {
+            return -EINVAL;
+        }
         if (sock != NULL) {
-            /* bind sock object implicitly */
             sock->local.port = src_port;
             if (remote == NULL) {
                 sock->local.family = sock->remote.family;
             }
             else {
                 sock->local.family = remote->family;
             }
+            gnrc_sock_create(&sock->reg, GNRC_NETTYPE_UDP, src_port);
 #ifdef MODULE_GNRC_SOCK_CHECK_REUSE
             /* prepend to current socks */
             sock->reg.next = (gnrc_sock_reg_t *)_udp_socks;
             _udp_socks = sock;
-#endif
-            gnrc_sock_create(&sock->reg, GNRC_NETTYPE_UDP, src_port);
+#endif /* MODULE_GNRC_SOCK_CHECK_REUSE */
         }
     }
     else {
         src_port = sock->local.port;
         memcpy(&local, &sock->local, sizeof(local));
     }
+    /* sock can't be NULL at this point */
     if (remote == NULL) {
-        /* sock can't be NULL at this point */
-        memcpy(&rem, &sock->remote, sizeof(rem));
+        rem = (sock_ip_ep_t *)&sock->remote;
         dst_port = sock->remote.port;
     }
     else {
-        memcpy(&rem, remote, sizeof(rem));
+        rem = (sock_ip_ep_t *)remote;
```

Comment:
At least it should go into a separate commit. And for my taste: more, but smaller PRs are **less** overhead.